### ISSUE TITLE
Issues with "Save to Blackhole" button when using SSL

### ISSUE THREAD

**Issue by ansuz07** [Jackett/Jackett#3793]
**Jackett version**:  0.10.164.0
**Mono version** (if not using Windows):  

-------------------------------

So, I have a very strange little problem.  I've got my Jackett server running via reverse proxy on IIS and overall it works well.  However, there is a small issue when working with the Save to Blackhole directory functionality when I'm browsing the secure site.  Here are a few scenarios:

- Browse the insecure site:  Blackhole functionality works correctly - the file saves to the correct location and I get the "success" message in Jackett
- Browse the secure site:  Blackhole functionality does not work - the file does not save and I get the "failure" message.  However, if I copy the link from the "save to blackhole" button and paste it directly into my browser bar, the functionality works correctly - the file saves to the correct location and I get the "success" message in Jackett.

So, in summary, the blackhole functionality works correctly, _except_ if I use the button on the secure site.

I've looked in the logs and I can't find any mention of the failure, even with enhanced logging turned on. 

Any ideas what is going on?

**Comment by flightlevel** [comment]
Can you share your IIS configuration

**Comment by ansuz07** [comment]
Sure.  It is a pretty standard reverse proxy in IIS, where jacket his hosted on another machine with IP:Port and IIS rewrites Jackett.domain.com to the port via the rule.  Beyond the web.config file below, all I've done is have IIS preserve host headers:

?xml version="1.0" encoding="UTF-8"?>
<configuration>
    <system.webServer>
        <rewrite>
            <rules>
                <rule name="HTTPS" enabled="true" patternSyntax="Wildcard" stopProcessing="true">
                    <match url="*" />
                    <conditions logicalGrouping="MatchAll" trackAllCaptures="false">
                        <add input="{HTTPS}" pattern="off" />
                    </conditions>
                    <action type="Redirect" url="https://{HTTP_HOST}{REQUEST_URI}" redirectType="Found" />
                </rule>
		<rule name="ReverseProxyInboundRule1" stopProcessing="true">
                    <match url="(.*)" />
                    <action type="Rewrite" url="http://IP:PORT/{R:1}" />
                    <serverVariables>
                        <set name="HTTP_X_ORIGINAL_ACCEPT_ENCODING" value="{HTTP_ACCEPT_ENCODING}" />
                        <set name="HTTP_ACCEPT_ENCODING" value="" />
                        <set name="HTTP_X_ORIGINAL_HOST" value="{HTTP_HOST}" />
                    </serverVariables>
                </rule>
            </rules>
            <outboundRules>
                <clear />
                <rule name="RestoreAcceptEncoding" preCondition="NeedsRestoringAcceptEncoding">
                    <match serverVariable="HTTP_ACCEPT_ENCODING" pattern="^(.*)" />
                    <conditions logicalGrouping="MatchAll" trackAllCaptures="true" />
                    <action type="Rewrite" value="HTTP_X_ORIGINAL_ACCEPT_ENCODING" />
                </rule>
                <rule name="ReverseProxyOutboundRule1" preCondition="ResponseIsHtml1">
                    <match filterByTags="A, Form, Img" pattern="^http(s)?://IP:Port/(.*)" />
                    <conditions logicalGrouping="MatchAll" trackAllCaptures="true" />
                    <action type="Rewrite" value="http{R:1}://jackett.domain.com/{R:2}" />
                </rule>
 		<rule name="Rewrite Location Header" preCondition="IsRedirection?" patternSyntax="ExactMatch">
                    <match serverVariable="RESPONSE_Location" pattern="^http://[^/]+/(.*)" />
                    <conditions>
                        <add input="{ORIGINAL_URL}" pattern=".+" />
                    </conditions>
                    <action type="Rewrite" value="http{R:1}://{ORIGINAL_URL}/{C:1}/{R:2}" />
                </rule>
                <preConditions>
                    <preCondition name="ResponseIsHtml1">
                        <add input="{RESPONSE_CONTENT_TYPE}" pattern="^text/(.+)" />
                    </preCondition>
                    <preCondition name="NeedsRestoringAcceptEncoding">
                        <add input="HTTP_X_ORIGINAL_ACCEPT_ENCODING" pattern=".+" />
                    </preCondition>
			<preCondition name="IsRedirection?">
                        <add input="{RESPONSE_STATUS}" pattern="3\d\d" />
                    </preCondition>
                </preConditions>
            </outboundRules>
        </rewrite>
    </system.webServer>
</configuration>

**Comment by kaso17** [comment]
I've no experince with IIS but it looks like the `X-Forwarded-Proto` header is not set (should be http/https)

**Comment by ansuz07** [comment]
Thank you so much for the help - setting the X-Forwarded-Proto header did the trick.

For anyone interested, here is the updated IIS config file:

     <?xml version="1.0" encoding="UTF-8"?>
<configuration>
    <system.webServer>
        <rewrite>
            <rules>
                <rule name="HTTPS" enabled="true" patternSyntax="Wildcard" stopProcessing="true">
                    <match url="*" />
                    <conditions logicalGrouping="MatchAll" trackAllCaptures="false">
                        <add input="{HTTPS}" pattern="off" />
                    </conditions>
                    <action type="Redirect" url="https://{HTTP_HOST}{REQUEST_URI}" redirectType="Found" />
                </rule>
		<rule name="ReverseProxyInboundRule1" stopProcessing="true">
                    <match url="(.*)" />
                    <action type="Rewrite" url="http://Jacket IP:  Jacket PORT /{R:1}" />
                    <serverVariables>
                        <set name="HTTP_X_ORIGINAL_ACCEPT_ENCODING" value="{HTTP_ACCEPT_ENCODING}" />
                        <set name="HTTP_ACCEPT_ENCODING" value="" />
                        <set name="HTTP_X_ORIGINAL_HOST" value="{HTTP_HOST}" />
                        <set name="HTTP_X_FORWARDED_HOST" value="{HTTP_HOST}" />
                        <set name="HTTP_X_FORWARDED_SCHEMA" value="https" />
                        <set name="HTTP_X_FORWARDED_PROTO" value="https" />
                    </serverVariables>
                </rule>
            </rules>
            <outboundRules>
                <clear />
                <rule name="RestoreAcceptEncoding" preCondition="NeedsRestoringAcceptEncoding">
                    <match serverVariable="HTTP_ACCEPT_ENCODING" pattern="^(.*)" />
                    <conditions logicalGrouping="MatchAll" trackAllCaptures="true" />
                    <action type="Rewrite" value="HTTP_X_ORIGINAL_ACCEPT_ENCODING" />
                </rule>
                <rule name="ReverseProxyOutboundRule1" preCondition="ResponseIsHtml1">
                    <match filterByTags="A, Form, Img" pattern="^http(s)?://Jacket IP:  Jacket PORT/(.*)" />
                    <conditions logicalGrouping="MatchAll" trackAllCaptures="true" />
                    <action type="Rewrite" value="http{R:1}://FQDN/{R:2}" />
                </rule>
 		<rule name="Rewrite Location Header" preCondition="IsRedirection?" patternSyntax="ExactMatch">
                    <match serverVariable="RESPONSE_Location" pattern="^http://[^/]+/(.*)" />
                    <conditions>
                        <add input="{ORIGINAL_URL}" pattern=".+" />
                    </conditions>
                    <action type="Rewrite" value="http{R:1}://{ORIGINAL_URL}/{C:1}/{R:2}" />
                </rule>
                <preConditions>
                    <preCondition name="ResponseIsHtml1">
                        <add input="{RESPONSE_CONTENT_TYPE}" pattern="^text/(.+)" />
                    </preCondition>
                    <preCondition name="NeedsRestoringAcceptEncoding">
                        <add input="HTTP_X_ORIGINAL_ACCEPT_ENCODING" pattern=".+" />
                    </preCondition>
			<preCondition name="IsRedirection?">
                        <add input="{RESPONSE_STATUS}" pattern="3\d\d" />
                    </preCondition>
                </preConditions>
            </outboundRules>
        </rewrite>
    </system.webServer>
</configuration>

**Comment by kaso17** [comment]
Is there some simple way of making HTTP_X_FORWARDED_PROTO dynamic (based on the original requests's protocol)?

**Comment by ansuz07** [comment]
Honestly, I have no idea.  I'm an IIS novice, so its just a guess, but I would think you could build it into a separate rule where it only rewrites the variable if the HTTPS value is "on".  I use a similar trick in the first rule to rewrite everything HTTP to HTTPS if HTTPS is "off" (the rule doesn't fire if the traffic is already HTTPS) so I'd imaging you could create a similar rule to only rewrite the header when HTTP is "on"

I'll play around with it tonight and see if I can make that work.

**Comment by kaso17** [comment]
Thank you, if you could try to build a minimal working configuration which we could add to the readme that would be appreciated.
On the wiki there's a IIS example but that probably will have the same problem as you experienced.
https://github.com/Jackett/Jackett/wiki/Reverse-Proxy

**Comment by ansuz07** [comment]
Sure thing.  Give me a couple of days and I'll write up a working config for IIS

**Comment by ansuz07** [comment]
So, I got bored and typed it all up.  I've tested these configurations on my own IIS server and they work perfectly.

[IIS Jackett Configuration.docx](https://github.com/Jackett/Jackett/files/2389499/IIS.Jackett.Configuration.docx)
